src/main/java/com/tcs/edu/decorator/OrderDoublingProcessor.java

```diff
@@ -0,0 +1,93 @@
+package com.tcs.edu.decorator;
+
+import com.tcs.edu.MessageProcessor;
+import com.tcs.edu.domain.Doubling;
+import com.tcs.edu.domain.Message;
+import com.tcs.edu.domain.MessageOrder;
+import com.tcs.edu.service.ValidatingService;
+
+/**
+ * Процессинг массива сообщений в соответствии с заданным порядком и наличием дублей
+ *
+ * @author  t.m.kharchenko
+ * @see #process(MessageOrder, Message...)              Метод для сортировки массива сообщений по заданному порядку
+ * @see #process(MessageOrder, Doubling, Message...)    Метод для исключения дублей в массиве сообщений по заданному порядку
+ */
+
+public class OrderDoublingProcessor extends ValidatingService implements MessageProcessor {
+
+    /**
+     * Метод для сортировки массива сообщений по заданному порядку
+     * @param order                 Перечислимый тип (переменная типа MessageOrder) с порядком сортировки
+     * @param messages              Массив сообщений (переменная типа Message[])
+     * @return                      Массив сообщений
+     * @see OrderDoublingProcessor  Родительский класс
+     */
+    @Override
+    public Message[] process (MessageOrder order, Message... messages) {
+        super.isArgsValid(order);
+        Message[] processedMessages = new Message[messages.length];
+        switch (order) {
+            case ASC: {
+                processedMessages = messages;
+                break;
+            }
+            case DESC: {
+                int currentDescMessage = 0;
+                for (int currentMessage = messages.length - 1; currentMessage >= 0; currentMessage--) {
+                    processedMessages[currentDescMessage] = messages[currentMessage];
+                    currentDescMessage++;
+                }
+                break;
+            }
+        }
+        return processedMessages;
+    }
+
+    /**
+     * Перегруженный метод для определения дублей в отсортированном массиве сообщений
+     * @param order                 Перечислимый тип (переменная типа MessageOrder) с порядком сортировки
+     * @param doubling              Перечислимый тип (переменная типа Doubling) с определением наличия дублей
+     * @param messages              Массив сообщений (переменная типа Message[])
+     * @return                      Массив сообщений
+     * @see OrderDoublingProcessor  Родительский класс
+     */
+    @Override
+    public Message[] process(MessageOrder order, Doubling doubling, Message... messages) {
+        super.isArgsValid(order, doubling);
+        Message[] processedMessages = null;
+        switch (doubling) {
+            case DOUBLES: {
+                processedMessages = process (order, messages);
+                break;
+            }
+            case DISTINCT: {
+                int currentMessageIndex = 0;
+                int distinctWrittenMessageIndex = 0;
+                Message[] processedMessagesWithEmpties = new Message[messages.length];
```

Comment:
Массивы это конечно хорошо, но реально работать с ними неудобно, поэтому если есть возможность используй подходящую коллекцию, например:
```suggestion
                final List<Message> distinctMessages = new ArrayList<>();
                for (Message message : messages) {
                    if (!distinctMessages.contains(message)) {
                        distinctMessages.add(message);
                    }
                }
                return distinctMessages.toArray(new Message[messages.length]);
```